runtime/job-workers/client-functions.cpp

```diff
@@ -86,65 +105,127 @@ Optional<int64_t> f$kphp_job_worker_start(const class_instance<C$KphpJobWorkerRe
+      first = false;
     } else {
-      memory_manager.release_shared_message(memory_request);
-      critical_section.leave_critical_section();
-      php_warning("Can't send job: probably jobs queue is full");
-      return false;
+      if (cur_shared_mem_piece.get() != common_shared_memory_piece.get()) {
+        php_warning("Can't extract common shared memory piece in multiple jobs request: requests[%s] and requests[%s] have different shared memory pieces",
+                    requests.begin().get_key().to_string().c_str(), it.get_key().to_string().c_str());
+        return {};
+      }
     }
   }
 
-  int64_t job_resumable_id = register_forked_resumable(new job_resumable{job_id});
+  array<Optional<int64_t>> res{requests.size()};
+
+  job_workers::JobSharedMemoryPiece *common_job_request = nullptr;
+  if (!common_shared_memory_piece.is_null()) {
+    common_job_request = make_job_request_message<job_workers::JobSharedMemoryPiece>(common_shared_memory_piece);
+    /**
+     * common_job_request lifetime:
+     * 1. attaches to client process on creating in `acquire_shared_message()`
+     * 2. attaches to job worker process on job receiving in `job_parse_execute()`
+     * 3. detaches from job worker process on terminating in `release_shared_message()` recursively
+     * 4. detaches from client process at the end of this function in `release_shared_message()` recursively
+     *
+     * client: 1 -> 4
+     *         ↓
+     * server: 2 -> 3
+     */
+    if (common_job_request == nullptr) {
+      return {};
+    }
+  }
 
-  update_precise_now();
-  kphp_event_timer *timer = allocate_event_timer(get_precise_now() + timeout, get_job_timeout_wakeup_id(), job_id);
+  for (const auto &it : requests) {
+    const auto &req = it.get_value();
 
-  vk::singleton<job_workers::ProcessingJobs>::get().start_job_processing(job_id, job_workers::JobRequestInfo{job_resumable_id, timer});
-  return job_resumable_id;
-}
+    req.get()->set_shared_memory_piece({});                                                 // prepare for copying to shared memory
+    auto *job_request = make_job_request_message<job_workers::JobSharedMessage>(req); // copy to shared memory
+    req.get()->set_shared_memory_piece(common_shared_memory_piece);                         // roll it back to keep original instance unchanged
+    if (job_request == nullptr) {
+      res.set_value(it.get_key(), false);
+      continue;
+    }
 
-class_instance<C$KphpJobWorkerResponse> f$kphp_job_worker_wait(int64_t job_resumable_id, double timeout) noexcept {
-  if (!f$is_kphp_job_workers_enabled()) {
-    php_warning("Can't wait job: job workers disabled");
-    return {};
+    if (common_job_request) {
+      const auto &job_instance = job_request->instance.cast_to<C$KphpJobWorkerRequest>();
+      const auto &common_job_instance = common_job_request->instance.cast_to<C$KphpJobWorkerSharedMemoryPiece>();
```

Comment:
let's save `common_job_instance` by value and move it later
`job_instance.get()->set_shared_memory_piece(std::move(common_job_instance));`